docs/developer/api-versioning.md

```diff
@@ -0,0 +1,391 @@
+
+If the server is upgraded and some old supported versions are phased
+out, the client may be caught by surprise.
+
+Servers should respond with status `404`, error label
+`unsupported-version`.  The versioning middleware can do that (see
+above).
+
+Client should install a catch-all that will handle this specific
+error, re-fetch `/api-versions`, and try again.
+
+This will only happen if backend and client grow apart in time beyond
+the supported limit, and there is some chance it will result in an
+empty set of compatible versions, so it's also ok to just fail here.
+
+
+## Strongly typed versions
+
+If we make version an ADT `ApiVersion`, we can remove old versions
+from it in one place and have the compiler guide us through all the
+places where we need to remove it.
+
+There are at least two ways to implement this:
+
+1. Add a few extra servant instances for `(v :: ApiVersion) :> route`.
+2. Define a type family `Versions` that maps `V*` to `"v*"`, and write
+`Versions version :>` in the routing type instead of `"v1"`.
+
+2 seems a lot less work to write, read, and understand.
+
+
+## Data migration (aka data marshalling)
+
+If the shape of an end-point changes between versions (if a data type
+in the routing table becomes a type family), it is often possible to
+write marshalling functions that translate a value from an older
+version into one of a newer version or vice versa.
+
+These functions are called marshalling functions and are useful to
+define separately to keep the application logic clean.
+
+For certain changes to a data type used in an API, marshalling is
+straight-forward in both directions.  The most common example is
+adding an optional attribute to a JSON object:
+
+- *backward migration*: remove the new attribute.
+- *forward migration*: set the new attribute to `null`.
+
+(This is what wire has traditionally done to accomplish client
+backwards compatibility without any API versioning.)
+
+If a mandatory attribute is added in a newer version, there may be a
+plausible default value that can be used in the forward migration
+(backward migration would still remove the field).
+
+In other cases, whether there is an automatic migration depends on the
+use case and the semantics.
+
+It may even be impossible to marshal either in one or in both
+directions.  In this case, you have 3 options:
+
+1. abandon compatibility;
+2. rethink your new version and craft it in a way that two-way
+   marshalling is possible;
+3. make the application work around the gap, eg. by gracefully
+   refusing to offer video conferencing in a client if it is not
+   supported on the server yet.
+
+
+## Writing client code
+
+If you write all code by hand and don't generate anything from the
+swagger docs, just look at the swagger diff for every new version and
+take it from there.
+
+If you generate, say, typescript or kotlin or swift from swagger:
+
+0. have a generated source module `Gen.ts`, plus a source module with
```

Comment:
We will version a library per version of the API achieving something similar to what you describe I believe

Reply:
That may end you up with more dead code in the executable, but that may not be a problem.  Implementation detail!